book/emails.rst

```diff
@@ -0,0 +1,320 @@
+            return array('email');
+        }
+
+        public function getId()
+        {
+            return $this->id;
+        }
+
+        public function getEmails()
+        {
+            return $this->emails;
+        }
+
+        public function getFirstName()
+        {
+            return $this->firstName;
+        }
+
+        public function getLastName()
+        {
+            return $this->lastName;
+        }
+    }
+
+.. _book-emails-owner-provider:
+
+Implementing the ``EmailOwnerProviderInterface``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to make the application able to find the owner of a certain e-mail address, you have to
+create a provider that implements the
+:class:`Oro\\Bundle\\EmailBundle\\Entity\\Provider\\EmailOwnerProviderInterface`. This interface
+contains two methods:
+
+``getEmailOwnerClass()``
+    This is the class of the e-mail owner entity (the class implementing the
+    ``EmailOwnerInterface`` which is the ``Applicant`` class in the example above).
+
+``findEmailOwner()``
+    Returns an entity that is the owner of an e-mail address or ``null`` if no such owner exists.
+    The returned object must be an instance of the class specified by ``getEmailOwnerClass()``.
+
+The provider class should then look like this:
+
+.. code-block:: php
+    :linenos:
+
+    // src/Acme/DemoBundle/Entity/Provider/EmailOwnerProvider.php
+    namespace Acme\DemoBundle\Entity\Provider;
+
+    use Acme\DemoBundle\Entity\Applicant;
+    use Acme\DemoBundle\Entity\ApplicantEmail;
+    use Doctrine\ORM\EntityManager;
+    use Oro\Bundle\EmailBundle\Entity\Provider\EmailOwnerProviderInterface;
+
+    class EmailOwnerProvider implements EmailOwnerProviderInterface
+    {
+        public function getEmailOwnerClass()
+        {
+            return 'Oro\Bundle\UserBundle\Entity\User';
+        }
+
+        public function findEmailOwner(EntityManager $em, $email)
+        {
+
+            $applicantRepository = $em->getRepository('AcmeDemoBundle:Applicant');
+            $applicant = $applicantRepository->findOneBy(array('email' => $email));
+
+            if (null === $applicant) {
+                $emailRepository = $em->getRepository('AcmeDemoBundle:Email');
+                $emailEntity = $emailRepository->findOneBy(array('email' => $email));
+
+                if (null !== $emailEntity) {
+                    $applicant = $emailEntity->getEmailOwner();
+                }
+            }
+
+            return $applicant;
+        }
+    }
+
+You then need to create a service for the new ``EmailOwnerProvider`` class and tag it with the
+``oro_email.owner.provider`` tag to make the application aware of the new e-mail provider:
+
+.. code-block:: yaml
+    :linenos:
+
+    # src/Acme/DemoBundle/Resources/config/services.yml
+    services:
+        acme_demo:
+            class: Acme\DemoBundle\Entity\Provider\EmailOwnerProvider
+            tags:
+                - { name: oro_email.owner.provider, order: 1 }
```

Comment:
Maybe it is better to specify more higher order value, for example `3` - to put newly created provider at the end of providers chain